Use hierarchical structure for serialzed datasets

sonicaj opened the pull request:
This commit updates datasets_serialized return output to give datasets in hierarchical structure only as flattened/hybrid is just having duplicates of the same data and the consumer in this case can take care of the flattening process.

william-gr reviewed (commented):
Why is the copy necessary?

sonicaj committed:
Use hierarchical structure for serialzed datasets

This commit updates datasets_serialized return output to give datasets in hierarchical structure only as flattened/hybrid is just having duplicates of the same data and the consumer in this case can take care of the flattening process.

sonicaj committed:
Add mountpoint key by default

This commit adds mountpoint key by default to the datasets serialized output. However a knob has been exposed where if this is not desired it can be explicitly mentioned.

sonicaj committed:
Allow retrieving no property for better speed

This commit exposes a knob which allows consumers to not retrieve any prop which is way faster then when we retrieve them.

sonicaj committed:
Expose knob for retrieving custom prop

This commit exposes knobs to control where custom user defined props should be retrieved or not.

william-gr commented on libzfs.pyx:
```diff
@@ -549,11 +549,13 @@ cdef class ZFS(object):
         prop_mapping = {}
+        # If props is None, we include all properties, if it's an empty list, no property is retrieved
+        # except for mountpoint which is separate
         for prop_id in ZFS.proptypes[DatasetType.FILESYSTEM]:
```
How about `for prop_id in (ZFS.proptypes[DatasetType.FILESYSTEM] if props is None or len(props) != 0 else []):`

then we avoid even iterating if we dont want any props.

william-gr commented on libzfs.pyx:
```diff
@@ -535,34 +533,27 @@ cdef class ZFS(object):
-        return fields_found
-
-    def datasets_serialized(self, props=None):
+    def datasets_serialized(self, props=None, mountpoint=True, custom_props=True):
```
Why mountpoint is special? That seems weird

sonicaj committed:
Allow setting props in top level dict

This commit updates serialized datasets api to allow setting different props to be in the top level dict when the data is retrieved. One thing to note here is that if a key is specified to be in the top level dict and hasn't been specified by props attribute, it will raise a value error and also keeping in line with legacy implementation, we add mountpoint key by default.
